### PR TITLE
Versioning of plans

### LINKED ISSUE #157: Add versioning to plans

### LINKED ISSUE #164: Add versioning of upvotes and comments

### PR DESCRIPTION
**Pull request by keanecjy**
At PlansView level:
* Listen to all plans with that tripId
* Update with latest versioned plan of each plan only

At PlanView level:
* Listen to all plans with that planId
* Update with latest versioned plan only

Closes #157 
Closes #164

### CODE REVIEW

**Comment by Vielheim** [comment]
Versioning bug:
- When we update a plan to a newer version, any comments or upvotes reference to the initial version of the view
- E.g. we start at version 10, and edit until version 13. the comments and upvotes at ver.13 will reference version 10

**Comment by keanecjy** [comment]
I think comments can just fetch all though? I thought we wanted the full history for comments

**Comment by Vielheim** [comment]
Yes, but the model still references some version number. I think the issue is when planViewModel updates, it is not propagated to the commentsViewModel or upvoteViewModel

**Comment by keanecjy** [comment]
yep, think we can drill the props down, but not too sure how it works

**Comment by Vielheim** [comment]
Not sure if it is a good idea, but maybe we can make CommentsVM and UpvoteVM a delegate of PlanViewModel, then when the plan is updated, we call the delegates to retrieve the updated planID

So it'll be an array of delegates in PlanVM.

**Comment by Vielheim** [comment]
I added a sample for Comments, seems to work, but not sure about its correctness

**Comment by Vielheim** [comment]
When two users are editing the same plan version, they can both submit and the plan version will only increment by 1. Feature/bug?

^ Maybe can use a delegate to update EditPlanViewModel as well?

**Comment by Vielheim** [comment]
Maybe editing plan VM can also use a delegate? For: #166

**Comment by Vielheim** [comment]
Added a picker for version, doesn't actually do anything yet. Just a draft

**Comment by Vielheim** [comment]
- Added `allDays` property for version control
- Now PlansVM and PlanVM holds a delegate for each plan they own. Any changes to the VM will be published to the delegates (mainly to update their version numbers)

File: TourMate/TourMate/Frontend/Plan/Views/PlansViews/PlansViewModel.swift [hunk 1]
```diff
@@ -116,6 +132,11 @@ extension PlansViewModel {
         }
 
         self.plans = Array(latestPlanMap.values)
+
+        for plan in self.plans {
+            await planEventDelegates[plan.id]?.update(plan: plan, errorMessage: "")
+        }
```
Comment (keanecjy) [inline TourMate/TourMate/Frontend/Plan/Views/PlansViews/PlansViewModel.swift:138]: I think we have to assign the plan event delegate to newly added plans?

**Comment by keanecjy** [comment]
Looks good! I think we can merge

**Review by Vielheim — APPROVED** [review]

**Pull request merged**